Mathlib/LinearAlgebra/Multilinear/Curry.lean

```diff
@@ -223,87 +223,114 @@ def multilinearCurryRightEquiv :
 
 namespace MultilinearMap
 
-variable {ι' : Type*} {R M₂}
 
-/-- A multilinear map on `∀ i : ι ⊕ ι', M'` defines a multilinear map on `∀ i : ι, M'`
-taking values in the space of multilinear maps on `∀ i : ι', M'`. -/
-def currySum (f : MultilinearMap R (fun _ : ι ⊕ ι' => M') M₂) :
-    MultilinearMap R (fun _ : ι => M') (MultilinearMap R (fun _ : ι' => M') M₂) where
+variable {R M₂} {ι' : Type*} {N : (ι ⊕ ι') → Type*}
+  [∀ i, AddCommMonoid (N i)] [∀ i, Module R (N i)]
+
+/-- Given a family of modules `N : (ι ⊕ ι') → Type*`, a multilinear map
+on `(fun _ : ι ⊕ ι' => M')` induces a multilinear map on
+`(fun (i : ι) ↦ N (.inl i))` taking values in the space of
+linear maps on `(fun (i : ι') ↦ N (.inr i))`. -/
+def currySum (f : MultilinearMap R N M₂) :
+    MultilinearMap R (fun (i : ι) ↦ N (.inl i))
+      (MultilinearMap R (fun (i : ι') ↦ N (.inr i)) M₂) where
   toFun u :=
-    { toFun := fun v => f (Sum.elim u v)
-      map_update_add' := fun v i x y => by
-        letI := Classical.decEq ι
-        simp only [← Sum.update_elim_inr, f.map_update_add]
-      map_update_smul' := fun v i c x => by
-        letI := Classical.decEq ι
-        simp only [← Sum.update_elim_inr, f.map_update_smul] }
+    { toFun := fun v ↦ f (Sum.rec u v)
+      map_update_add' := fun _ _ _ _ ↦ by letI := Classical.decEq ι; simp
+      map_update_smul' := fun _ _ _ _ ↦ by letI := Classical.decEq ι; simp }
   map_update_add' u i x y :=
-    ext fun v => by
-      letI := Classical.decEq ι'
-      simp only [MultilinearMap.coe_mk, add_apply, ← Sum.update_elim_inl, f.map_update_add]
+    ext fun _ ↦ by letI := Classical.decEq ι'; simp
   map_update_smul' u i c x :=
-    ext fun v => by
-      letI := Classical.decEq ι'
-      simp only [MultilinearMap.coe_mk, smul_apply, ← Sum.update_elim_inl, f.map_update_smul]
+    ext fun _ ↦ by letI := Classical.decEq ι'; simp
 
 @[simp]
-theorem currySum_apply (f : MultilinearMap R (fun _ : ι ⊕ ι' => M') M₂) (u : ι → M')
-    (v : ι' → M') : f.currySum u v = f (Sum.elim u v) :=
```

Comment:
I think we should keep around an `elim` version, since `Sum.elim` is fairly common and not `simp`-ed to `Sum.rec`.